Report shard counts for ongoing snapshots

arteam opened the pull request:
For in-progress snapshots we can try to calculate the amount of successful shards based on the shard state. We can use a similar approach for failures.

Reference #76704

arteam committed:
Report shard counts for ongoing snapshots

Ref #76704

Add simple unit tests for shard counts

arteam commented:
@elasticsearchmachine 
run elasticsearch-ci/part-1
run elasticsearch-ci/rest-compatibility

arteam commented:
@elasticsearchmachine update branch

arteam commented:
@elasticsearchmachine
run elasticsearch-ci/part-1
run elasticsearch-ci/rest-compatibility

arteam commented:
@elasticmachine update branch

arteam commented:
@elasticmachine run elasticsearch-ci/part-1

arteam commented:
@elasticmachine update branch

arteam commented:
@elasticmachine update branch

arteam commented:
@elasticmachine update branch

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam committed:
Calculate only successful shards

arteam commented:
@elasticmachine update branch

arteam commented:
@elasticmachine update branch

arteam commented:
@elasticmachine run elasticsearch-ci/part-1

arteam commented:
@elasticmachine run elasticsearch-ci/part-1

arteam committed:
Don't check stable pagination for GetSnapshotsIT

arteam committed:
Update snapshotStatus

arteam commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:
```diff
@@ -178,10 +178,6 @@ public void testSortAndPaginateWithInProgress() throws Exception {
         awaitNumberOfSnapshotsInProgress(inProgressCount);
 
         final String[] repos = { repoName };
-        assertStablePagination(repos, allSnapshotNames, GetSnapshotsRequest.SortBy.START_TIME);
```
`assertStablePagination` fails for in-progress snapshots, because we get a different amount of `successfulShards` for different calls, so getting all the data and getting it page by page doesn't yield the same result. I'm not sure if there's a way pinpoint the amount of successfully migrated shards. Currently, we only block the data nodes

DaveCTurner replied:
The issue is that some of the shards are unchanged so don't need to write any data files and therefore we don't block them from succeeding. I think we can just wait for them all to succeed:

```diff
diff --git a/server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java b/server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java
index 03fc2ad0529..d5a832f0079 100644
--- a/server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java
@@ -14,6 +14,7 @@ import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRes
 import org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsRequest;
 import org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsRequestBuilder;
 import org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsResponse;
+import org.elasticsearch.cluster.SnapshotsInProgress;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.search.sort.SortOrder;
 import org.elasticsearch.threadpool.ThreadPool;
@@ -176,8 +177,22 @@ public class GetSnapshotsIT extends AbstractSnapshotIntegTestCase {
             inProgressSnapshots.add(startFullSnapshot(repoName, snapshotName));
         }
         awaitNumberOfSnapshotsInProgress(inProgressCount);
+        awaitClusterState(
+            state -> state.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY)
+                .entries()
+                .stream()
+                .flatMap(s -> s.shards().stream())
+                .allMatch(
+                    e -> e.getKey().getIndexName().equals("test-index-1") == false
+                        || e.getValue().state() == SnapshotsInProgress.ShardState.SUCCESS
+                )
+        );

         final String[] repos = { repoName };
+        assertStablePagination(repos, allSnapshotNames, GetSnapshotsRequest.SortBy.START_TIME);
+        assertStablePagination(repos, allSnapshotNames, GetSnapshotsRequest.SortBy.NAME);
+        assertStablePagination(repos, allSnapshotNames, GetSnapshotsRequest.SortBy.INDICES);
+
         assertThat(
             clusterAdmin().prepareGetSnapshots(matchAllPattern())
                 .setSnapshots(GetSnapshotsRequest.CURRENT_SNAPSHOT, "-snap*")
```

arteam replied:
Awesome!

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam committed:
Make sure we wait until all shards have been snapshot

Otherwise, the stable sorting isn't working because we get a different amount
of the successfully migrated shards.

arteam committed:
RUn spotless:apply

arteam commented:
Hi @tlrx and @original-brownbear, could you guys please take a look at the PR, thanks!

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into report-in-progress-snapshots-stats

arteam committed:
Randomize the amount of shards for snapshots in progress

tlrx reviewed (approved):
LGTM

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java:
```diff
@@ -426,7 +427,13 @@ public void testGetSnapshotsWithSnapshotInProgress() throws Exception {
+        SnapshotInfo snapshotInfo = snapshotInfoList.get(0);
+        assertEquals(SnapshotState.IN_PROGRESS, snapshotInfo.state());
+
+        SnapshotStatus snapshotStatus = client().admin().cluster().prepareSnapshotStatus().execute().actionGet().getSnapshots().get(0);
```
Let's use `get` instead of `actionGet()` please so we don't swallow stack traces that make understanding failures a pain :)

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotInfo.java:
```diff
@@ -379,9 +380,13 @@ public SnapshotInfo(SnapshotsInProgress.Entry entry) {
-            0,
-            Collections.emptyList(),
+            entry.shards().size(),
+            Math.toIntExact(entry.shards().stream().filter(c -> c.getValue().state() == SnapshotsInProgress.ShardState.SUCCESS).count()),
```
Might be nice to calculate both this count and the failures list in a single loop by using a static method + `Tuple` or so? Afaik this may run on a transport thread so it's worth saving the cycles, especially as we scale to larger and larger clusters.

arteam replied:
Yeah, I believe I can do that with a `teeing` collector.

arteam replied:
Actually, we can't, because we were calling another constructor and couldn't declare variables. I went ahead with removing of the constructor call and using a simple loop for calculating both fields.

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -1207,6 +1207,10 @@ public void testSnapshotStatus() throws Exception {
         assertThat(getResponse.getSnapshots().size(), equalTo(1));
         SnapshotInfo snapshotInfo = getResponse.getSnapshots().get(0);
         assertThat(snapshotInfo.state(), equalTo(SnapshotState.IN_PROGRESS));
+        snapshotStatus = client.admin().cluster().prepareSnapshotStatus().execute().actionGet().getSnapshots().get(0);
```
Lets use `get` here as well

original-brownbear reviewed (commented):
Looks good, just a few minor points tests (lets not use `actionGet`) and the looping.

arteam committed:
Calculate successfulShards and shardFailures in one go

arteam commented:
@elasticmachine run elasticsearch-ci/part-1

arteam commented:
Thanks Tanguy and Armin!

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 77622`
